### PR TITLE
I1492 - edits made while getting bb8 setup on support

### PR DESCRIPTION
**Pull request by hillalex**

### CODE REVIEW

File: starport_setup_steps_taken.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ssh annex.montagu
+git clone https://github.com/vimc/bb8
+sudo useradd bb8 -U -d /var/lib/bb8
+sudo mkdir -p /var/lib/bb8/
+sudo chown -R bb8:bb8 /var/lib/bb8
+sudo su bb8
+cd ~
+mkdir .ssh
+/home/martin/bb8/setup_starport.sh
+mkdir starport
```
Comment (hillalex) [inline starport_setup_steps_taken.sh]: One missing step - `adduser bb8 shh`

Comment (hillalex) [reply starport_setup_steps_taken.sh]: Once I added `bb8` to the `ssh` group the below worked successfully!

File: setup.sh [hunk 2]
```diff
@@ -24,6 +24,7 @@ fi
 
 # give bb8 group ownership of this dir
 chgrp -R bb8 $HERE
+chmod 775 $HERE
```
Comment (hillalex) [inline setup.sh]: Make sure group has write permissions to directory

File: starport_setup_steps_taken.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# Starport
```
Comment (hillalex) [inline starport_setup_steps_taken.sh]: I'll remove this file and add to the README instead

File: setup.sh [hunk 4]
```diff
@@ -17,13 +17,14 @@ if ! id -u bb8 > /dev/null 2>&1; then
...
     mkdir -p /var/lib/bb8/.ssh
-
+    
     # add to docker group
     usermod -aG docker bb8
 fi
 
 # give bb8 group ownership of this dir
 chgrp -R bb8 $HERE
+chmod -R 775 $HERE
```
Comment (hillalex) [inline setup.sh]: Needed to make sure members of `bb8` group can write to the directory

File: configs/support/config.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        "backup_location": "~bb8/starport/"
+    },
+    "targets": [
+        {
+          "name": "registry_data",
+          "type": "named_volume"
+        },
+        {
+            "type": "directory",
+            "path": "/montagu/vault/storage"
```
Comment (richfitz) [inline configs/support/config.json:13]: Why do we not have vault using a volume?  Would be a really simple change in the vault setup and brings it further in line with the rest of the containerised bits

File: schedule.py [hunk 6]
```diff
@@ -27,7 +27,7 @@ def add_job(cron, backup_script, test_job):
 def schedule_backups(test_job):
     here = dirname(abspath(__file__))
     print("Scheduling backup task")
-    backup_script = join(here, "backup.py")
+    backup_script = "cd {} && {}".format(here, join(here, "backup.sh"))
```
Comment (richfitz) [inline schedule.py]: I would prefer it if the cd logic was in `backup.sh` because otherwise every time we call it we need to make sure we're in the right directory

File: setup.sh (hunk 4 above)
Comment (richfitz) [reply setup.sh]: Is the -R needed?  That will set the executable bit on all files.  Might be better to do `chmod -R g+w $HERE` instead?  (sorry for the misdirection)

File: README.md [hunk 7]
```diff
@@ -56,6 +56,63 @@ There are five entrypoints to the backup module.
...
-4. `schedule.py`: Creates a cron job that runs backup.py every hour and logs to `./log`. Must be run as root.
+4. `schedule.py`: Creates a cron job that runs backup.sh every hour and logs to `./log`. Must be run as root.
 5. `restore.py`:  TODO
 
+# Setup
+These are the steps followed to set up starport on the annex:
+1.  Add a bb8 user and group
+
+```
+sudo useradd bb8 -U -d /var/lib/bb8
```
Comment (richfitz) [inline README.md]: I am forgetting things, but somewhere else, in another bb8 setup script, the setup script itself set up the bb8 user.  Can we not do the same here?

File: README.md [hunk 8]
```diff
@@ -56,6 +56,63 @@ There are five entrypoints to the backup module.
...
+```
+
+4. Clone this
+```
+git clone https://github.com/vimc/bb8
+```
+
+5. Create a `.ssh` directory for the `bb8` user
+```
+mkdir .ssh
```
Comment (richfitz) [inline README.md]: why is this needed?

File: README.md [hunk 9]
```diff
@@ -56,6 +56,63 @@ There are five entrypoints to the backup module.
...
+```
+
+6. Create the starport directory as the `bb8` user
```
Comment (richfitz) [inline README.md]: all this user stuff makes me really glad we have gone with docker elsewhere!  But also makes me wonder if this should all be running as docker too so that we can do all the setup as just an image and then do the backup into a further volume?

**Review by richfitz — CHANGES_REQUESTED** [review]
Bunch of small things, but the big one that I think would warrant further discussion is:

> all this user stuff makes me really glad we have gone with docker elsewhere! But also makes me wonder if this should all be running as docker too so that we can do all the setup as just an image and then do the backup into a further volume?

You and Martin may have already discussed this?  But given how much permissions have already bitten us, I wonder if this would be a worthwhile thing to think about

File: README.md (hunk 7 above)
Comment (hillalex) [reply README.md]: Yeah, it was designed so the program itself has to run as `bb8`, but the backup location is configurable. I hadn't actually envisioned having a bb8 user on the starport - originally I was connecting as the existing montagu user, meaning no extra setup was needed on the annex. @MartinEden added the bb8 user on the annex, which required these extra steps. 

Either we could leave the backup location configurable, and just add an extra script that installs a bb8 user that can be used if desired, or we could enforce having a bb8 user on the remote machine, i.e. actually add it in the `setup_starport` script (in which case should remove the configurable backup options from the config file.) Which do you think is preferable?

File: README.md (hunk 8 above)
Comment (hillalex) [reply README.md]: So we can add the private key as an authorized key and connect via ssh as the bb8 user

Comment (richfitz) [reply README.md]: where does that happen? Can we make the directory at the point we add the keys?

File: setup.sh (hunk 4 above)
Comment (hillalex) [reply setup.sh]: So does `chmod -R g+w $HERE` give write permissions for the group to the whole directory and subdirectories?

**Comment by MartinEden** [comment]
With regards to permissions/docker, I think we should do one of two things:

* Insist that we use a bb8 user on all machines, and make the `setup_starport.sh` and `setup.sh` scripts create them, rather than there being manual steps.
* Do everything in Docker (I'm a bit vague on this)

I think this middle ground of doing things on metal but leaving user management up to the specific case is too muddy.

File: configs/support/config.json (hunk 5 above)
Comment (hillalex) [reply configs/support/config.json:13]: Sure, lets open a ticket for this

File: setup.sh (hunk 4 above)
Comment (richfitz) [reply setup.sh]: That's right - so rather than setting all permissions to `rwxrwxr-x` which is what the 775 would do, this adds the `w` bit for the group - I should have thought about this when we discussed the other day

File: README.md (hunk 8 above)
Comment (hillalex) [reply README.md]: If we're moving all the bb8 user setup to the `setup_starport` script then yes

**Comment by hillalex** [comment]
Ok, this now has all the user permissioning logic wrapped up in the setup scripts. Both need to be run as `sudo`.

**Review by MartinEden — APPROVED** [review]
Looks good to me :)

**Comment by richfitz** [comment]
Cool - @MartinEden and I had a quick chat about dockerising this and I'd like to follow up with you about this Monday perhaps Alex?

**Review by richfitz — APPROVED** [review]

**Comment by hillalex** [comment]
Sure, lets chat about docker in person on Mon
